x/validator-preference/README.md

```diff
@@ -0,0 +1,148 @@
+# Validator Set Preference 
+
+## Abstract 
+
+Validator-Set preference is a new module which gives users and contracts a 
+better UX for staking to a set of validators. For example: a one click button
+that stakes to multiple validators. Then the user can set (or realistically a frontend provides) 
+a list of recommended defaults (Ex: governoors, wosmongton, chain/stack contributors etc).
+Currently this can be done on-chain with frontends, but having a preference list stored locally 
+eases frontend code burden. 
+
+## Design 
+
+How does this module work? 
+
+- Allow an user to set a list of {val-addr, weight} in state, called their validator-set preference.
+- Give users a single message to stake {X} tokens, according to their validator set preference distribution.
+- Give users a single message to unstake {X} tokens, according to their validator set preference distribution.
+- Give users a single message to claim rewards from everyone on their preference list.
+- If a user does not have a validator set preference list set, and has staked do the following; 
+  - Make their preference list default to their current staking distribution.
+- If a user has no preference list and no staking, then return error for messages.
+
+## Calculations
+
+Staking Calculation 
+
+- The user provides an amount to stake and our `MsgStakeToValidatorSet` divides the amount based on validator weight distribution.
+  For ex: Stake 100osmo with validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2}
+  our stake logic will attempt to stake (100 * 0.5) 50osmo for ValA , (100 * 0.3) 30osmo from ValB and (100 * 0.2) 20osmo from ValC.
+
+UnStaking Calculation 
+
+- The user provides an amount to unstake and our `MsgUnStakeFromValidatorSet` divides the amount based on validator weight distribution.
```

Comment:
What if I stake with set A, then I change my preferences to different proportions (set B) and try to unstake? Is there some kind of history of preferences to make sure that we don't attempt to unstake more or fewer tokens based on the changed preference list?

Reply:
hmm thats interesting. Are you recommending something like this approach? 

```
UserA -> {ValA: 0.5, ValB: 0.2, ValC: 0.3}
User A stakes: 100osmo {ValA-> 50osmo, ValB-> 20osmo, ValC -> 30osmo} 
** Store the history of {weight, amount} ** 

UserA modifies the weight -> {ValA: 0.1, ValB: 0.2, ValC: 0.8}
** Internally check if this modification is possible (i.e. check if the user staked >= amount to that validator) **  
** Only allow weight modification if the newAmountToStake < oldAmountStaked which can be calculated by weight*TotalAmount ** 

UserA unstakes 50osmo -> {ValA -> 5osmo, ValB -> 10osmo, ValC -> 40osmo} 
**Result in error because ValC only has 30tokens staked** 
```

Reply:
On another look, I think I just misinterpreted what is happening here. 

I guess when userA changes their preferences list, the unstaking from the old set and restaking to a new set is going to happen behind the scenes. That way, the problem I originally mentioned wouldn't happen. It might be worth specifying in the doc that changing the preferences list implies that we unstake from the old + restake to a new set.

Reply:
aah i see. Yea i believe this is a much cleaner approach. I will mention it :))

Reply:
I'm going to leave this unresolved, so that we can get more feedback from other reviewers

Reply:
So this spec doesn't discuss the case of what is done when a user does some 'direct' undelegation.

So I set valset { A: .5, B: .2, C: .3 }, and suppose I stake 100 to it, to get:

A: 50, B: 20, C: 30

Then I directly undelegate 20 from B. (So I'm at A:50, C:30)

If I try to valset undelegate 80, what should happen. Options that I think are fine:
- valset Undelegate fails
- valset Undelegate Takes {arbitrary distribution across valset validators if theres been direct modification that is unsatisfiable}

Reply:
I find this generally fine, since redelegate to valset will fix this in worst case scenarios

Reply:
hmm i didn't really think about "direct unstake" and "val-set undelegate" colliding. Should i look into updating "validator-set" state based on the txs from "direct undelegate" and vice versa?